File: library/src/main/java/com/thesurix/gesturerecycler/EmptyViewDataObserver.java
package com.thesurix.gesturerecycler;

import android.support.v7.widget.RecyclerView;
import android.view.View;

/**
 * Observer class for managing visibility of the adapter's empty view.
 */
class EmptyViewDataObserver extends RecyclerView.AdapterDataObserver {

    private RecyclerView mRecyclerView;
    private View mEmptyView;

    @Override
    public void onChanged() {
        updateEmptyViewState();
    }

    @Override
    public void onItemRangeInserted(final int positionStart, final int itemCount) {
        updateEmptyViewState();
    }

    @Override
    public void onItemRangeRemoved(final int positionStart, final int itemCount) {
        updateEmptyViewState();
    }

    public void setEmptyView(final View emptyView) {
        mEmptyView = emptyView;
        updateEmptyViewState();
    }

    public void setRecyclerView(final RecyclerView recyclerView) {
        mRecyclerView = recyclerView;
        updateEmptyViewState();
    }

    private void updateEmptyViewState() {
        if (mEmptyView != null && mRecyclerView != null) {
            if (mRecyclerView.getAdapter().getItemCount() == 0) {
                mEmptyView.setVisibility(View.VISIBLE);
                mRecyclerView.setVisibility(View.GONE);
            } else {
                mEmptyView.setVisibility(View.GONE);
                mRecyclerView.setVisibility(View.VISIBLE);
            }
        }
    }
}

File: library/src/main/java/com/thesurix/gesturerecycler/EmptyViewDataObserver.kt
package com.thesurix.gesturerecycler

import android.view.View
import androidx.recyclerview.widget.RecyclerView

/**
 * Observer class for managing visibility of the adapter's empty view.
 */
internal class EmptyViewDataObserver : RecyclerView.AdapterDataObserver() {

    var recyclerView: RecyclerView? = null
        set(value) {
            field = value
            updateEmptyViewState()
        }
    var emptyView: View? = null
        set(value) {
            field = value
            updateEmptyViewState()
        }

    var emptyViewVisibilityListener: EmptyViewVisibilityListener? = null

    override fun onChanged() {
        updateEmptyViewState()
    }

    override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
        updateEmptyViewState()
    }

    override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
        updateEmptyViewState()
    }

    private fun updateEmptyViewState() {
        recyclerView?.let {
            val noItems = it.adapter?.itemCount == 0
            emptyView?.visibility = if (noItems) View.VISIBLE else View.GONE
            recyclerView?.visibility = if (noItems) View.GONE else View.VISIBLE
            emptyViewVisibilityListener?.onVisibilityChanged(noItems)
        }
    }
}

interface EmptyViewVisibilityListener {
    fun onVisibilityChanged(visible: Boolean)
}
